### ISSUE TITLE
Build failure on NetBSD

### ISSUE THREAD

**Issue by enkl7** [Arisotura/melonDS#1016]
Build fails on both ARM64 and x86 on NetBSD 9.1

ARM64:

```
[ 56%] Building CXX object src/CMakeFiles/core.dir/ARMJIT_Memory.cpp.o
/home/x/melonDS/src/ARMJIT_Memory.cpp: In function ‘void SigsegvHandler(int, siginfo_t*, void*)’:
/home/x/melonDS/src/ARMJIT_Memory.cpp:173:60: error: ‘struct mcontext_t’ has no member named ‘fault_address’
         desc.EmulatedFaultAddr = (u8*)context->uc_mcontext.fault_address - curArea;
                                                            ^~~~~~~~~~~~~
/home/x/melonDS/src/ARMJIT_Memory.cpp:174:50: error: ‘struct mcontext_t’ has no member named ‘pc’
         desc.FaultPC = (u8*)context->uc_mcontext.pc;
                                                  ^~
/home/x/melonDS/src/ARMJIT_Memory.cpp:194:34: error: ‘struct mcontext_t’ has no member named ‘pc’
             context->uc_mcontext.pc = (u64)desc.FaultPC;
                                  ^~
*** Error code 1
```

I'm able to bypass this error by adding an ifdef:
```
#elif defined(__NetBSD__)
    _UC_MACHINE_SET_PC(context, (u64)desc.FaultPC);
```

However, now I'm stuck at the linking process which is where it stops on x86 as well with the same error:
```

[ 67%] Building ASM object src/CMakeFiles/core.dir/ARMJIT_A64/ARMJIT_Linkage.S.o
[ 68%] Linking CXX static library libcore.a
Error running link command: No such file or directory
*** Error code 2
```

**Comment by WaluigiWare64** [comment]
The error is due to LTO. We set the `CMAKE_AR` and `CMAKE_RANLIB` to `gcc-ar` and `gcc-ranlib`.
...however, they don't seem to exist on NetBSD.
In the [NetBSD Docs](https://man.netbsd.org/gcc.1) they mention:
```
           If object files containing GIMPLE bytecode are stored in a library
           archive, say libfoo.a, it is possible to extract and use them in an
           LTO link if you are using a linker with plugin support.  To create
           static libraries suitable for LTO, use gcc-ar and gcc-ranlib
           instead of ar and ranlib; to show the symbols of object files with
           GIMPLE bytecode, use gcc-nm.  Those commands require that ar,
           ranlib and nm have been compiled with plugin support.  At link
           time, use the the flag -fuse-linker-plugin to ensure that the
           library participates in the LTO optimization process.
```
Perhaps the default gcc version is too old?
(I built with clang and lld by the way.)

**Comment by WaluigiWare64** [comment]
Just built gcc10 from pkgsrc.
It has `gcc-ar` and `gcc-ranlib` in `/usr/pkg/gcc10/bin`, and I can confirm that setting this in CMakeLists.txt makes it build fine.

**Comment by enkl7** [comment]
@WaluigiWare64 
Thanks for the quick response! 
Yes, the default gcc version on NetBSD turned out to be too old, I successfully compiled melonDS with both clang and gcc10. With this, x86 compiles with no problem. 

ARM64 however still needs an `#ifdef` patch to successfully compile even with LTO disabled `-DENEBLE_LTO=OFF`.

Nevertheless, closing this issue since I managed to successfully build on both x86 and ARM64.

**Comment by WaluigiWare64** [comment]
We shouldn't close this issue - i still need to fix NetBSD ARM64.

**Comment by WaluigiWare64** [comment]
Could you try building on ARM64 NetBSD from the [fastmem-pc branch](https://github.com/Arisotura/melonDS/tree/fastmem-pc)?

**Comment by enkl7** [comment]
@WaluigiWare64 Building the fastmem-pc branch on ARM64 NetBSD worked without a problem. Thank you so much!

Unfortunately, there's barely any performance difference enabling the JIT recompiler on ARM64 NetBSD and the current performance is unplayable. On the contrary, it's a night and day difference on ARM64 Linux with smooth performance when enabling JIT - I'm using a RockPRO64 and Pinebook Pro. I guess this is for a new issue?

**Comment by WaluigiWare64** [comment]
> 
> 
> @WaluigiWare64 Building the fastmem-pc branch on ARM64 NetBSD worked without a problem. Thank you so much!
> 
> Unfortunately, there's barely any performance difference enabling the JIT recompiler on ARM64 NetBSD and the current performance is unplayable. On the contrary, it's a night and day difference on ARM64 Linux with smooth performance when enabling JIT - I'm using a RockPRO64 and Pinebook Pro. I guess this is for a new issue?

Could you try using software renderer (if you aren't already) @enkl7 ?

**Comment by WaluigiWare64** [comment]
(I've also merged that branch into master.)

**Comment by enkl7** [comment]
@WaluigiWare64 Yes, already tried using the software renderer and enabling/disabling the optimization options as well as changing the JIT block size etc. The same problem regarding performance is similar on x86 (tested on Optiplex 3020M)

I have been testing with different emulator settings, nda roms and different compilers to try to figure out more about the bad performance (low frame rate, stuttering and audio distortion etc):

**ARM64**
gcc 7.5 - minor graphical glitches
clang 10.0.1 - a lot of graphical glitches with lines across screen.
gcc10 - almost 0 graphical glitches (not noticeable unless you really try to find them)

Difference in compiler did not affect frame rate, I was able to squeeze a few extra fps by tweaking JIT block size and changing the value of hw.audio0.blk_ms with sysctl. These are very minor performance tweaks, the play experience is still not enjoyable but it's now at least at a _playable_ frame rate in comparison to the older melonDS version in pkgsrc.

**x86**
Same glitches depending on which compiler is used as mentioned above on ARM64.

The strange thing is that performance with JIT enabled is about the same (overall smooth frame rate with some occasional stuttering) as the older melonDS version already available in pkgsrc that doesn't have JIT supported. Turning off JIT I would expect the same performance as the older melonDS but it performs worse with slightly more audio stuttering and occasional frame drops.
